dali/kernels/reduce/find_reduce.cuh

```diff
@@ -0,0 +1,118 @@
+#include "dali/core/geom/vec.h"
+
+#include "dali/kernels/reduce/reduce_drop_dims.h"
+#include "dali/kernels/reduce/reduce_gpu_impl.cuh"
+
+namespace dali {
+namespace kernels {
+
+/**
+ * @brief Preprocess step that returns the index of the sample if the value satisfies the predicate
+ *        The neutral value (the "not found" value) will depend on the reduction that we need to apply.
+ */
+template <typename T, typename Predicate, typename Reduction, typename Acc = ptrdiff_t>
+struct FindReducePreprocess {
+  const Predicate *__restrict__ predicate = nullptr;
+  const T *__restrict__ base = nullptr;
+
+  DALI_HOST_DEV DALI_FORCEINLINE Acc operator()(const T &x) const noexcept {
+    if (!predicate || (*predicate)(x)) {
+      return &x - base;
+    } else {
+      return Reduction::template neutral<Acc>();
+    }
+  }
+};
+
+/**
+ * @brief Postprocess step required to convert the min/max neutral value to -1
+          (meaning "not found")
+ */
+template <typename Out, typename Reduction>
+struct FindReducePostprocess {
+  template <typename T>
+  DALI_HOST_DEV DALI_FORCEINLINE Out operator()(T x) const {
+    return x == Reduction::template neutral<T>() ? Out(-1) : x;
+  }
+};
+
+/**
+ * @brief Finds positions in the input that satisfy a predicate, reducing them
+ *        to a single index (e.g. first, last)
+ */
+template <typename Out, typename In, typename Predicate, typename Reduction,
+          typename Acc = ptrdiff_t>
+class FindReduceGPU
+    : public reduce_impl::ReduceImplGPU<Out, In, Acc,
+                                        FindReduceGPU<Out, In, Predicate, Reduction, Acc>> {
+ public:
+  using ReduceBase =
+      reduce_impl::ReduceImplGPU<Out, In, Acc, FindReduceGPU<Out, In, Predicate, Reduction, Acc>>;
+  using Preprocessor = FindReducePreprocess<In, Predicate, Reduction, Acc>;
+  using Postprocessor = FindReducePostprocess<Out, Reduction>;
+
+  std::vector<const In *> base_;
+  const Predicate *predicates_ = nullptr;
+
+  void InitSampleData(const InListGPU<In> &in) {
+    int nsamples = in.num_samples();
+    base_.resize(nsamples);
+    for (int i = 0; i < nsamples; i++) {
+      base_[i] = in[i].data;
+    }
+  }
+
+  void InitPredicates(const Predicate *predicates) {
+    predicates_ = predicates;
+  }
+
+  Preprocessor GetPreprocessorImpl(int sample_idx, bool batch) const {
+    assert(!batch);  // not allowed
+    return Preprocessor{predicates_ + sample_idx, base_[sample_idx]};
+  }
+
+  Postprocessor GetPostprocessorImpl(int sample_index, bool batch) const {
+    assert(!batch);  // not allowed
+    return {};
+  }
+
+  Reduction GetReduction() const {
+    return {};
+  }
+
+  void Setup(KernelContext &kctx, const TensorListShape<> &in_shape) {
+    std::array<int, 1> axes = {0};
+    ReduceBase::Setup(kctx, in_shape, make_cspan(axes), false, false);
+  }
+
+  void Run(KernelContext &kctx, const OutListGPU<Out> &out, const InListGPU<In> &in,
+           const Predicate *predicates = nullptr) {
```

Comment:
With this definition, the predicates must be in contiguous device memory. This limits the reusability of the kernels - imagine we expose it as an operator `find_first_last` with a threshold value, which may be on the GPU. In general, we should not assume that the inputs are contiguous - this would require coalescing - copying data to an intermediate buffer with a scatter/gather kernel. The alternative is to use a TensorListView of predicates.
```suggestion
           const InListGPU<Predicate, 0> predicates) {
```

Reply:
Good point. Done.